engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/aggregate/TableAggregator.scala

```diff
@@ -1,34 +1,234 @@
+    }
+
+  }
+
+  private val minMaxAllowedTypes = List(
+    TINYINT,
+    SMALLINT,
+    INTEGER,
+    BIGINT,
+    FLOAT,
+    DOUBLE,
+    BOOLEAN,
+    VARCHAR,
+    DECIMAL,
+    TIME_WITHOUT_TIME_ZONE,
+    DATE,
+    TIMESTAMP_WITHOUT_TIME_ZONE,
+    TIMESTAMP_WITH_LOCAL_TIME_ZONE
+  )
+
+  // As of Flink 1.19, time-related types are not supported in FIRST_VALUE aggregate function.
+  // See: https://issues.apache.org/jira/browse/FLINK-15867
+  // See AggFunctionFactory.createFirstValueAggFunction
+  private val firstLastAllowedTypes = List(
+    TINYINT,
+    SMALLINT,
+    INTEGER,
+    BIGINT,
+    FLOAT,
+    DOUBLE,
+    BOOLEAN,
+    VARCHAR,
+    DECIMAL
+  )
+
+  private val numericAggregationsAllowedTypes = List(
+    TINYINT,
+    SMALLINT,
+    INTEGER,
+    BIGINT,
+    FLOAT,
+    DOUBLE,
+    DECIMAL
+  )
+
 }
 
 sealed trait TableAggregator extends EnumEntry {
-  val name: String
-  val inputTypeConstraint: Option[TypingResult]
-  def outputType(inputTypeInRuntime: TypingResult): TypingResult
-  val flinkFunctionName: String
+
+  val displayName: String
+
+  def flinkFunctionDefinition: BuiltInFunctionDefinition
+
+  val flinkFunctionName: String = flinkFunctionDefinition.getName
+
+  def inputAllowedTypesConstraint: Option[List[LogicalTypeRoot]]
+
+  def inferOutputType(inputType: TypingResult)(
+      implicit nodeId: NodeId
+  ): Either[ProcessCompilationError, TypingResult] = {
+    def buildError()(implicit nodeId: NodeId) = CustomNodeError(
+      s"Invalid type: '${inputType.withoutValue.display}' for selected aggregator: '$displayName'.",
+      Some(aggregateByParamName)
+    )
+    def validateUsingInputTypeStrategy(callContext: NuAggregationFunctionCallContext) = {
+      flinkFunctionDefinition
+        .getTypeInference(DataTypeFactoryHolder.instance)
+        .getInputTypeStrategy
+        .inferInputTypes(callContext, false)
+        .toScala
+        .flatMap(_.asScala.headOption) match {
+        case Some(_) => Right(())
+        case None    => Left(buildError())
+      }
+    }
+    def validateUsingAllowedTypesConstraint() = {
+      this.inputAllowedTypesConstraint match {
+        case Some(allowedTypes) =>
+          if (inputType.toDataType.getLogicalType.isAnyOf(allowedTypes: _*)) {
+            Right(())
+          } else {
+            Left(buildError())
+          }
+        case None => Right(())
+      }
+    }
+    def inferOutputType(
+        callContext: NuAggregationFunctionCallContext
+    ) = {
+      flinkFunctionDefinition
+        .getTypeInference(DataTypeFactoryHolder.instance)
+        .getOutputTypeStrategy
+        .inferType(callContext)
+        .toScala match {
+        case Some(value) => Right(value.getLogicalType.toTypingResult)
+        case None        => Left(buildError())
+      }
+    }
+    val callContext = NuAggregationFunctionCallContext(inputType, flinkFunctionDefinition)
+    for {
+      _          <- validateUsingInputTypeStrategy(callContext)
```

Comment:
Let's wrap validateUsingInputTypeStrategy andThen inferOutputType in method called e.g. `validateAgainstFlinkBuiltInFunctionDefinition`